File: modulo03-mcp-na-pratica/mcp-customers-server-java/src/main/java/com/iadeva/mcp/service/CustomerService.java
package com.iadeva.mcp.service;

// Equivalente ao customerService.ts — camada de negócio entre MCP e HTTP client
// Adiciona funcionalidade de busca/filtro em memória que a API legada não oferece nativamente

import com.iadeva.mcp.client.CustomerHttpClient;
import com.iadeva.mcp.model.Customer;
import com.iadeva.mcp.model.MutationResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * Serviço de domínio para operações com clientes.
 * Delega chamadas HTTP ao {@link CustomerHttpClient} e adiciona lógica de negócio
 * como filtragem em memória por múltiplos critérios.
 */
@Service
public class CustomerService {

    private static final Logger logger = LoggerFactory.getLogger(CustomerService.class);

    private final CustomerHttpClient httpClient;

    public CustomerService(CustomerHttpClient httpClient) {
        this.httpClient = httpClient;
    }

    /**
     * Retorna todos os clientes cadastrados.
     *
     * @return lista completa de clientes
     */
    public List<Customer> listAll() {
        logger.debug("Listando todos os clientes");
        return httpClient.listAll();
    }

    /**
     * Busca e filtra clientes em memória por id, nome ou telefone.
     * Funcionalidade extra que a API legada não oferece — equivalente ao searchCustomer do TypeScript.
     * Todos os parâmetros são opcionais; se nenhum for fornecido retorna todos os clientes.
     * O filtro usa correspondência parcial case-insensitive para name e phone.
     *
     * @param id    filtra por ID exato (opcional)
     * @param name  filtra por nome contendo o valor (opcional, case-insensitive)
     * @param phone filtra por telefone contendo o valor (opcional)
     * @return lista de clientes que satisfazem todos os critérios fornecidos
     */
    public List<Customer> searchCustomer(String id, String name, String phone) {
        logger.debug("Buscando clientes — id={}, name={}, phone={}", id, name, phone);

        List<Customer> all = httpClient.listAll();

        return all.stream()
                .filter(c -> id == null || id.isBlank() || id.equals(c.id()))
                .filter(c -> name == null || name.isBlank()
                        || (c.name() != null && c.name().toLowerCase().contains(name.toLowerCase())))
                .filter(c -> phone == null || phone.isBlank()
                        || (c.phone() != null && c.phone().contains(phone)))
                .toList();
    }

    /**
     * Cria um novo cliente.
     *
     * @param name  nome do cliente
     * @param phone telefone do cliente
     * @return resultado da operação
     */
    public MutationResult createCustomer(String name, String phone) {
        logger.debug("Criando cliente — name={}", name);
        return httpClient.create(name, phone);
    }

    /**
     * Atualiza um cliente existente.
     *
     * @param id    identificador do cliente
     * @param name  novo nome (opcional)
     * @param phone novo telefone (opcional)
     * @return resultado da operação
     */
    public MutationResult updateCustomer(String id, String name, String phone) {
        logger.debug("Atualizando cliente — id={}", id);
        return httpClient.update(id, name, phone);
    }

    /**
     * Remove um cliente pelo ID.
     *
     * @param id identificador do cliente
     * @return resultado da operação
     */
    public MutationResult deleteCustomer(String id) {
        logger.debug("Removendo cliente — id={}", id);
        return httpClient.delete(id);
    }
}

File: modulo03-mcp-na-pratica/mcp-customers-server-go/service/customer_service.go
// Package service contém a camada de negócio entre o servidor MCP e o
// cliente HTTP — equivalente a CustomerService.java. Adiciona busca/filtro
// em memória que a API legada não oferece nativamente.
package service

import (
	"context"
	"strings"

	"mcp-customers-server/model"
)

// customerLister é a única operação do CustomerHttpClient de que este
// serviço depende — extraída como interface para permitir um duplo de teste
// simples (equivalente ao @Mock CustomerHttpClient em
// CustomerServiceTest.java, sem precisar de um framework de mocking).
type customerLister interface {
	ListAll(ctx context.Context) ([]model.Customer, error)
	FindByID(ctx context.Context, id string) *model.Customer
	Create(ctx context.Context, name, phone string) model.MutationResult
	Update(ctx context.Context, id, name, phone string) model.MutationResult
	Delete(ctx context.Context, id string) model.MutationResult
}

// CustomerService é o serviço de domínio para operações com clientes.
type CustomerService struct {
	httpClient customerLister
}

// NewCustomerService cria um CustomerService sobre o cliente HTTP informado.
func NewCustomerService(httpClient customerLister) *CustomerService {
	return &CustomerService{httpClient: httpClient}
}

// ListAll retorna todos os clientes cadastrados.
func (s *CustomerService) ListAll(ctx context.Context) ([]model.Customer, error) {
	return s.httpClient.ListAll(ctx)
}

// SearchCustomer busca e filtra clientes em memória por id, nome ou
// telefone. Todos os parâmetros são opcionais (string vazia = sem filtro); o
// filtro usa correspondência parcial case-insensitive para name e phone —
// equivalente a CustomerService.searchCustomer.
func (s *CustomerService) SearchCustomer(ctx context.Context, id, name, phone string) ([]model.Customer, error) {
	all, err := s.httpClient.ListAll(ctx)
	if err != nil {
		return nil, err
	}

	var result []model.Customer
	for _, c := range all {
		if id != "" && id != c.ID {
			continue
		}
		if name != "" && !strings.Contains(strings.ToLower(c.Name), strings.ToLower(name)) {
			continue
		}
		if phone != "" && !strings.Contains(c.Phone, phone) {
			continue
		}
		result = append(result, c)
	}
	return result, nil
}

// CreateCustomer cria um novo cliente.
func (s *CustomerService) CreateCustomer(ctx context.Context, name, phone string) model.MutationResult {
	return s.httpClient.Create(ctx, name, phone)
}

// UpdateCustomer atualiza um cliente existente.
func (s *CustomerService) UpdateCustomer(ctx context.Context, id, name, phone string) model.MutationResult {
	return s.httpClient.Update(ctx, id, name, phone)
}

// DeleteCustomer remove um cliente pelo ID.
func (s *CustomerService) DeleteCustomer(ctx context.Context, id string) model.MutationResult {
	return s.httpClient.Delete(ctx, id)
}
